packages/common/src/type-helpers.ts

```diff
@@ -73,3 +86,45 @@ export function errorCode(error: unknown): string | undefined {
 export function assertNever(msg: string, x: never): never {
   throw new TypeError(msg + ': ' + x);
 }
+
+/**
+ * According to the EcmaScript's spec, the default behavior of JavaScript's `x instanceof Y` operator is to walk up the
+ * prototype chain of object 'x', checking if any constructor in that hierarchy is _exactly the same object_ as the
+ * constructor function 'Y'.
+ *
+ * Unfortunately, it happens in various situations that different constructor function objects get created for what
+ * appears to be the very same class. This leads to surprising behavior where `instanceof` returns false though it is
+ * known that the object is indeed an instance of that class. One particular case where this happens is when constructor
+ * 'Y' belongs to a different realm than the constuctor with which 'x' was instantiated. Another case is when two copies
+ * of the same library gets loaded in the same realm.
+ *
+ * In practice, this tends to cause issues when crossing the workflow-sandboxing boundary (since Node's vm module
+ * really creates new execution realms), as well as when running tests using Jest (see https://github.com/jestjs/jest/issues/2549
+ * for some details on that one).
+ *
+ * This function injects a custom 'instanceof' handler into the prototype of 'clazz', which is both cross-realm safe and
+ * cross-copies-of-the-same-lib safe. It works by adding a special symbol property to the prototype of 'clazz', and then
+ * checking for the presence of that symbol.
+ */
+export function makeCrossRealmSafeInstanceOf(clazz: { prototype: any }, markerName: string): void {
```

Comment:
What happens if a user has both an older SDK version and 1.8.2?
Will instanceof checks still work or do we rely on the symbol?
Do we hit the `else` clause in line 118?

Reply:
Also, this could have been a class decorator.

Reply:
> What happens if a user has both an older SDK version and 1.8.2?
> Will instanceof checks still work or do we rely on the symbol?
> Do we hit the else clause in line 118?

Let say we have `theObject instanceof TheClass`. The `instanceof` handler is determined solely by the right hand side argument (ie. `TheClass`).

If `TheClass`'s module is <1.8.2, then the default `instanceof` algorithm will be used, no matter the version of the module from which `theObject` was instantiated.
- If `theObject` was instantiated using exactly `TheClass` constructor function object (ie. the module has the same file location and the same execution context), then `instanceof` will work as expected.
- In all other cases, the `instanceof` check is guaranteed to return false.
- That is the behavior we know from pre-1.8.0 releases.

If `TheClass`'s module is 1.8.2+, then, the symbol based check will be used, no matter the version of the module with which `theObject` was instantiated. Consequently:
- If `theObject` was instantiated using a pre-1.8.0 module, then the symbol based check will obviously fail; anyway `instanceof` would also fail in that situation, so no regression here.

  AFAICT the only possible regression that could results from this are the cases where we were previously checking an error's `name` rather than using the `instanceof` operator (for example, the worker shutdown throwable). It would be very easy to modify `makeCrossRealmSafeInstanceOf` to also allow name based comparisions, but I'm afraid that would only make things less predictable, as it would work when `TheClass`'s come from 1.8.2+ and `theObject` from any version, but wouldn't work when `theObject` is 1.8.2+, but `TheClass` is older.
- If `theObject` was instantiated using 1.8.0 or 1.8.1, then I expect that the symbol based `instanceof` check should work on classes that were marked with the symbol in 1.8.0; it would however fail on classes that were not marked in 1.8.0 but are now marked in 1.8.2.

  Once again, these would fail on the default `instanceof` algorithm anyway, so would not be a regression. Note however that the opposite (`TheClass` is 1.8.0, and `theObject` is 1.8.2+) wouldn't work. Therefore, mixing 1.8.0/1 with 1.8.2+ modules should definitively be avoided.
- If `theObject` was instantiated using 1.8.2+, then the symbol based `instanceof` handler should work correctly, even if `TheClass` and `theObject` are from different package versions.

> Do we hit the else clause in line 118?

In practice, the `else` clause should only ever be reached if a user extends one of our own class, on which we registered the symbol-based `instanceof`, then use that custom class as the right hand side argument of `instanceof`.

For example, lets imagine that the user does this:

```
 class MyCustomSubclassOfTemporalFailure extends TemporalFailure {
    // ...
 }
 const theObject = new MyCustomSubclassOfTemporalFailure();

 const isInstanceOf = (theObject instanceof MyCustomSubclassOfTemporalFailure);
```

Here, `MyCustomSubclassOfTemporalFailure`'s inherits the `@@hasInstance` function from `TemporalFailure`. Therefore, `instanceof` will implicitly call `TemporalFailure`'s `hasInstance`'s function, as `MyCustomSubclassOfTemporalFailure[@@hasInstance](theObject)`. That means that `this` will be equal to `MyCustomSubclassOfTemporalFailure`, but `clazz` will be `TemporalFailure`. We therefore fall into the `else` clause, and revert to the default behavior of `instanceof`, which works only in same-context cases.

Reply:
> Also, this could have been a class decorator.

Yeah, I was also considering that option, but was delaying because I want to make sure that it is safe to use in Node.js 14.18 to 20+ with the current version of TS. Digging into that right now.